### DISCUSSION TITLE
Accept stdin in .insert-string widget?

### DISCUSSION THREAD

**Discussion by mozirilla213** [akinomyoga/ble.sh#291] [Q&A] [answered]
I wanted to have a function that could be piped into and write the input to the next command line.

I tried:
```
> myprint() {
  [ "$1" = '-' ] && ble/widget/.insert-string "$(cat)" || ble/widget/.insert-string "$1"
}
# but
> echo "echo 'hello' | myprint -
>  # doesn't show on next command line
> myprint - <<< "echo hello"  # This does work
> echo hello
> myprint "echo hello"   # This also
> echo hello
```
Although the others options work, piping would be more intuitive and I don't understand why it doesn't work, the zsh version works fine:
```
> myprint() {
  [ "$1" = '-' ] && print -z  "$(cat)" || print -z "$1"
}
> echo "echo 'hello' | myprint -
> echo hello
```

Also, by the way, this widget is very useful for interactive scripts, couldn't it be renamed/wrapped to something simpler like `ble-print` so that users could find it more easily?, or maybe put it in the wiki.

**Comment by akinomyoga** [comment]
- The pipeline is run in subshells. If you would like to affect the context of the parent shell, you need to set `shopt -s lastpipe`.
- The functions whose basename starting with `.` is a private (internal) function of ble.sh and should not be directly called. You should use `ble/widget/insert-string`.
- But to begin with, `ble/widget/*` is not expected to be run in the user command. The above usage seems to work now, but I'm not sure if it wouldn't cause any other problems and if it would continue to work in the future.

**Comment by mozirilla213** [comment]
I see, well I wish there could be a user facing function that replaces the command line.

By the way, `shopt -s lastpipe` doesn't make it work for me
```
> shopt -s lastpipe
> myprint() { ble/widget/insert-string "$(cat)"; }
> echo "echo hello" | myprint
>
```

**Reply by akinomyoga** [reply]
Ah, you are right. Bash's `lastpipe` only works when the job control is turned off (`set +m`), but we usually use job controls in interactive shells.

I'll take this as a feature request for a new utility to insert a string into the next command line.

**Comment by akinomyoga** [comment, marked as answer]
In commit 2a524f3, I have implemented `ble append-line <string>`, which appends a line to the next command line.

This is an example:

```bash
$ ble append-line 'echo hello'[RET]
$ echo hello
```

This command can be used in subshells (which doesn't seem to be possible with Zsh's `print -z`). This command can be called multiple times to insert multiple lines.

```bash
$ (ble append-line 'echo hello');(ble append-line 'echo world')[RET]
$ echo hello
echo world
```
Of course, this can be used to insert the commands read from a pipe:


```bash
$ { echo echo hello; echo echo world; } | ble append-line "$(cat)"
$ echo hello
echo world
```

**Comment by mozirilla213** [comment]
Awesome work, I really like this feature!
